File: Java/medium/findDuplicateNumber.java
import java.util.*;

class Solution {
    public static int findDuplicate(int[] nums) {
        for(int num : nums){
            int index = num;
            if(index < 0)   index = -num;
            if(nums[index] < 0) return index;
            nums[index] *= -1;
        }

        return nums[0];
    }
    
    // Test: {1,1}
    // Duplicate Number = 1

    // Test: {1,2,1}       
    // Duplicate Number = 1

    // Test: {1,3,4,2,2}   
    // Duplicate Number = 2

    // Test: {3,1,3,4,2}   
    // Duplicate Number = 3
    
    public static void main(String[] args){
        List<int[]> testSet = Arrays.asList(
            new int[]{1,1},
            new int[]{1,2,1},
            new int[]{1,3,4,2,2},
            new int[]{3,1,3,4,2}
        );

        for(int[] test : testSet){
            System.out.print("Test: {");
            for(int i = 0; i < test.length; i++){
                System.out.print(test[i] + (i == (test.length - 1)? "}\n":","));
            }
            System.out.println("Duplicate Number = " + findDuplicate(test) + '\n');
        }
    }
}

File: CPP/medium/findDuplicateNumber.cpp
#include <vector>
#include <iostream>

int findDuplicate(std::vector<int>& nums) {
    int ans = nums[0];
    for(int num: nums){
        int index = num;
        if(index < 0){
            index = -index;
        }
        if(nums[index] < 0) return index;
        nums[index] *= -1;
    }
    return ans;
}

// nums = {1,1}
// Duplicate Number = 1

// nums = {1,2,1}
// Duplicate Number = 1

// nums = {1,3,4,2,2}
// Duplicate Number = 2

// nums = {3,1,3,4,2}
// Duplicate Number = 3

int main(){
    std::vector<std::vector<int>> testSet = {
        {1,1},
        {1,2,1},
        {1,3,4,2,2},
        {3,1,3,4,2}
    };

    for(std::vector<int> test : testSet){
        std::cout << "nums = {";
        for(int i = 0; i < test.size(); i++){
            std::cout << test[i] << (i == (test.size() - 1)? "}\n":",");
        }
        std::cout << "Duplicate Number = "<< findDuplicate(test)<< std::endl << std::endl;
    }
    return 0;
}